File: 09_exceptions/Exceptions.cs
using System;

namespace BasicCS.Chapter09
{
    internal class Exceptions
    {
        static void Main()
        {
            // 기본 try-catch-finally
            try
            {
                Console.Write("Enter a number: ");
                string input = Console.ReadLine()!;
                int number = int.Parse(input);
                Console.WriteLine($"You entered: {number}");

                int result = 100 / number;
                Console.WriteLine($"100 / {number} = {result}");
            }
            catch (FormatException ex)
            {
                Console.WriteLine($"Invalid format: {ex.Message}");
            }
            catch (DivideByZeroException)
            {
                Console.WriteLine("Cannot divide by zero");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Unexpected error: {ex.GetType().Name}: {ex.Message}");
            }
            finally
            {
                Console.WriteLine("Finally block always executes");
            }

            // throw 키워드
            try
            {
                ValidateAge(-5);
            }
            catch (ArgumentException ex)
            {
                Console.WriteLine($"Validation error: {ex.Message}");
            }

            // 예외 필터 (Exception filter, C# 6.0+)
            try
            {
                throw new InvalidOperationException("Custom error code 42");
            }
            catch (InvalidOperationException ex) when (ex.Message.Contains("42"))
            {
                Console.WriteLine($"Caught filtered exception: {ex.Message}");
            }

            // using 문 (안전한 리소스 관리)
            try
            {
                using var resource = new Resource("MyResource");
                resource.DoWork();
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error with resource: {ex.Message}");
            }

            // 사용자 정의 예외 (Custom exception)
            try
            {
                Withdraw(100, 200);
            }
            catch (InsufficientFundsException ex)
            {
                Console.WriteLine($"Custom exception: {ex.Message}, Balance: {ex.Balance:C}");
            }
        }

        static void ValidateAge(int age)
        {
            if (age < 0)
                throw new ArgumentException("Age cannot be negative", nameof(age));
            Console.WriteLine($"Age {age} is valid");
        }

        static void Withdraw(decimal balance, decimal amount)
        {
            if (amount > balance)
                throw new InsufficientFundsException(balance, amount);
            Console.WriteLine($"Withdrew {amount:C}");
        }
    }

    // 사용자 정의 예외 클래스 (Custom exception class)
    class InsufficientFundsException : Exception
    {
        public decimal Balance { get; }
        public decimal Amount { get; }

        public InsufficientFundsException(decimal balance, decimal amount)
            : base($"Insufficient funds. Balance: {balance:C}, Required: {amount:C}")
        {
            Balance = balance;
            Amount = amount;
        }
    }

    // IDisposable 리소스 예제
    class Resource : IDisposable
    {
        private readonly string _name;
        public Resource(string name) => _name = name;

        public void DoWork() => Console.WriteLine($"Resource '{_name}' doing work");

        public void Dispose() => Console.WriteLine($"Resource '{_name}' disposed");
    }
}

File: VB/10-exceptions/Exceptions.vb
Imports System
Imports System.IO

Module Program
    Sub Main()
        ' 기본 예외 처리
        Try
            Dim a As Integer = 10
            Dim b As Integer = 0
            Dim result = a \ b
            Console.WriteLine(result)
        Catch ex As DivideByZeroException
            Console.WriteLine($"0으로 나눌 수 없습니다: {ex.Message}")
        Catch ex As Exception
            Console.WriteLine($"일반 예외: {ex.Message}")
        Finally
            Console.WriteLine("Finally 블록은 항상 실행됩니다.")
        End Try

        ' Catch When
        Try
            Console.Write("숫자를 입력하세요: ")
            Dim input As String = Console.ReadLine()
            Dim num As Integer = Integer.Parse(input)
            Console.WriteLine($"입력값: {num}")
        Catch ex As FormatException When input.Length = 0
            Console.WriteLine("입력값이 비어있습니다.")
        Catch ex As FormatException
            Console.WriteLine("올바른 숫자 형식이 아닙니다.")
        Catch ex As OverflowException
            Console.WriteLine("숫자가 너무 크거나 작습니다.")
        End Try

        ' Throw 예제
        Try
            ValidateAge(-5)
        Catch ex As ArgumentException
            Console.WriteLine($"유효성 검사 오류: {ex.Message}")
        End Try

        ' 사용자 정의 예외
        Try
            Throw New CustomException("사용자 정의 예외 발생!", 1001)
        Catch ex As CustomException
            Console.WriteLine($"[{ex.ErrorCode}] {ex.Message}")
        End Try

        ' Using 문 (자원 자동 해제)
        Using writer As New StreamWriter("test.txt")
            writer.WriteLine("Hello, Using!")
        End Using
        Console.WriteLine("파일이 자동으로 닫혔습니다.")
    End Sub

    Sub ValidateAge(age As Integer)
        If age < 0 Then
            Throw New ArgumentException("나이는 음수일 수 없습니다.", NameOf(age))
        End If
    End Sub
End Module

' 사용자 정의 예외
Public Class CustomException
    Inherits Exception

    Public Property ErrorCode As Integer

    Public Sub New(message As String, errorCode As Integer)
        MyBase.New(message)
        Me.ErrorCode = errorCode
    End Sub
End Class
